src/backends/sysfs.rs

```diff
@@ -0,0 +1,1374 @@
+    use crate::vdo::Pd3p2ProductTypeVdo;
+    use crate::vdo::Pd3p2ProductVdo;
+    use crate::BcdWrapper;
+    use crate::BitReader;
+    use crate::Error;
+    use crate::FromBytes;
+    use crate::Result;
+
+    use super::SYSFS_TYPEC_PATH;
+
+    /// A mockable sysfs reader.
+    pub struct SysfsReader(Option<PathBuf>);
+
+    #[cfg_attr(test, automock)]
+    impl SysfsReader {
+        pub fn new() -> Result<Self> {
+            Ok(Self(None))
+        }
+
+        pub fn set_path(&mut self, path: &str) -> Result<()> {
+            self.0 = Some(super::check_path(path)?);
+            Ok(())
+        }
+
+        fn read_file(&mut self) -> Result<String> {
+            let path = self.0.take().expect("Path not set");
+            let string = std::fs::read_to_string(path)?;
+            Ok(string)
+        }
+
+        pub fn read_bcd(&mut self) -> Result<BcdWrapper> {
+            let content = self.read_file()?;
+            let mut chars = content.chars();
+
+            let high = chars
+                .next()
+                .ok_or(io::Error::new(io::ErrorKind::InvalidData, "File is empty"))?;
+            let _ = chars.next().ok_or(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "File is too short",
+            ))?;
+
+            // Sometimes we get simply "2"
+            let low = chars.next().unwrap_or('0');
+
+            let high = high
+                .to_digit(10)
+                .ok_or(io::Error::new(io::ErrorKind::InvalidData, "Invalid digit"))?;
+            let low = low
+                .to_digit(10)
+                .ok_or(io::Error::new(io::ErrorKind::InvalidData, "Invalid digit"))?;
+
+            let bcd = (high << 8) | low;
+
+            Ok(BcdWrapper(bcd))
+        }
+
+        pub fn read_opr(&mut self) -> Result<ConnectorCapabilityOperationMode> {
+            let content = self.read_file()?;
+            if content.contains("source") {
+                if content.contains("sink") {
+                    Ok(ConnectorCapabilityOperationMode::Drp)
+                } else {
+                    Ok(ConnectorCapabilityOperationMode::RpOnly)
+                }
+            } else {
+                Ok(ConnectorCapabilityOperationMode::RdOnly)
+            }
+        }
+
+        pub fn read_pd_revision(&mut self) -> Result<u32> {
+            let content = self.read_file()?;
+            let mut chars = content.chars();
+
+            let b0 = chars.next().ok_or(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "File is too short",
+            ))?;
+            let _ = chars.next().ok_or(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "File is too short",
+            ))?;
+            let b2 = chars.next().ok_or(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "File is too short",
+            ))?;
+
+            let rev = ((b0 as u32 - b'0' as u32) << 4) | (b2 as u32 - b'0' as u32);
```

Comment:
Isn't this doing similar logic to `read_bcd` above? Is there a reason this uses char-level operations instead of `to_digit` then?